e2e-tests/src/steps/application_install.rs

```diff
@@ -127,24 +130,50 @@ impl ApplicationSource {
             Self::LocalFile(path) => meroctl.application_install(node_name, path).await,
             Self::Url(url) => {
                 let response = reqwest::get(url).await?;
-                let bytes = response.bytes().await?;
-
-                let decoded_bytes = if url.ends_with(".gz") {
-                    let mut decoder = GzDecoder::new(&bytes[..]);
-                    let mut decompressed = Vec::new();
-                    decoder.read_to_end(&mut decompressed)?;
-                    decompressed
-                } else {
-                    bytes.to_vec()
-                };
-
                 let temp_path = "/tmp/app.wasm";
-                fs::write(&temp_path, decoded_bytes).await?;
+                let file = File::create(&temp_path).await?;
+
+                if url.ends_with(".gz") {
+                    let stream = response.bytes_stream();
+                    let (tx, rx) = mpsc::channel::<Result<Bytes, std::io::Error>>(1024);
+
+                    let write_to_file = tokio::spawn(async move {
+                        let stream = tokio_stream::wrappers::ReceiverStream::new(rx);
+                        let reader = StreamReader::new(stream);
+                        let mut decoder = GzipDecoder::new(BufReader::new(reader));
+                        let mut file = file;
+                        io::copy(&mut decoder, &mut file).await?;
+                        file.flush().await?;
+                        Ok::<_, io::Error>(())
+                    });
+
+                    let process_input = tokio::spawn(async move {
+                        let mut stream = stream;
+                        while let Some(chunk) = stream.next().await {
+                            let chunk =
+                                chunk.map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+                            tx.send(Ok(chunk))
+                                .await
+                                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+                        }
+                        Ok::<_, io::Error>(())
+                    });
```

Comment:
we shouldn't need to `spawn` or use a `channel` here, try [`split`](https://docs.rs/futures-util/latest/futures_util/io/trait.AsyncReadExt.html#method.split)-ting the `GzipDecoder` and joining two [`copy`](https://docs.rs/futures-util/latest/futures_util/io/fn.copy.html) tasks

```
task 1 := [ response       -> gzip write half ]
task 2 := [ gzip read half -> file            ]
```